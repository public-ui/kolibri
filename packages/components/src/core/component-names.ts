export let KolAbbrTag = 'kol-abbr' as const;
export let KolAccordionTag = 'kol-accordion' as const;
export let KolAlertTag = 'kol-alert' as const;
export let KolAlertWcTag = 'kol-alert-wc' as const;
export let KolAvatarTag = 'kol-avatar' as const;
export let KolAvatarWcTag = 'kol-avatar-wc' as const;
export let KolBadgeTag = 'kol-badge' as const;
export let KolBreadcrumbTag = 'kol-breadcrumb' as const;
export let KolButtonGroupTag = 'kol-button-group' as const;
export let KolButtonGroupWcTag = 'kol-button-group-wc' as const;
export let KolButtonLinkTag = 'kol-button-link' as const;
export let KolButtonTag = 'kol-button' as const;
export let KolButtonWcTag = 'kol-button-wc' as const;
export let KolCardTag = 'kol-card' as const;
export let KolDetailsTag = 'kol-details' as const;
export let KolDrawerTag = 'kol-drawer' as const;
export let KolFormTag = 'kol-form' as const;
export let KolHeadingTag = 'kol-heading' as const;
export let KolHeadingWcTag = 'kol-heading-wc' as const;
export let KolIconTag = 'kol-icon' as const;
export let KolImageTag = 'kol-image' as const;
export let KolIndentedTextTag = 'kol-indented-text' as const;
export let KolIndentedTextWcTag = 'kol-indented-text-wc' as const;
export let KolInputCheckboxTag = 'kol-input-checkbox' as const;
export let KolInputColorTag = 'kol-input-color' as const;
export let KolInputDateTag = 'kol-input-date' as const;
export let KolInputEmailTag = 'kol-input-email' as const;
export let KolInputFileTag = 'kol-input-file' as const;
export let KolInputNumberTag = 'kol-input-number' as const;
export let KolInputPasswordTag = 'kol-input-password' as const;
export let KolInputRadioTag = 'kol-input-radio' as const;
export let KolInputRangeTag = 'kol-input-range' as const;
export let KolInputTag = 'kol-input' as const;
export let KolInputTextTag = 'kol-input-text' as const;
export let KolKolibriTag = 'kol-kolibri' as const;
export let KolLinkButtonTag = 'kol-link-button' as const;
export let KolLinkGroupTag = 'kol-link-group' as const;
export let KolLinkTag = 'kol-link' as const;
export let KolLinkWcTag = 'kol-link-wc' as const;
export let KolLogoTag = 'kol-logo' as const;
export let KolModalTag = 'kol-modal' as const;
export let KolNavTag = 'kol-nav' as const;
export let KolPaginationTag = 'kol-pagination' as const;
export let KolPopoverWcTag = 'kol-popover-wc' as const;
export let KolProgressTag = 'kol-progress' as const;
export let KolQuoteTag = 'kol-quote' as const;
export let KolSelectTag = 'kol-select' as const;
export let KolSkipNavTag = 'kol-skip-nav' as const;
export let KolSpanWcTag = 'kol-span-wc' as const;
export let KolSpinTag = 'kol-spin' as const;
export let KolSplitButtonTag = 'kol-split-button' as const;
export let KolSymbolTag = 'kol-symbol' as const;
export let KolTableStatefulTag = 'kol-table-stateful';
export let KolTableStatelessTag = 'kol-table-stateless' as const;
export let KolTableStatelessWcTag = 'kol-table-stateless-wc' as const;
export let KolTableTag = 'kol-table' as const;
export let KolTabsTag = 'kol-tabs' as const;
export let KolTextareaTag = 'kol-textarea' as const;
export let KolToastContainerTag = 'kol-toast-container' as const;
export let KolToolbarTag = 'kol-toolbar' as const;
export let KolTooltipWcTag = 'kol-tooltip-wc' as const;
export let KolTreeItemTag = 'kol-tree-item' as const;
export let KolTreeItemWcTag = 'kol-tree-item-wc' as const;
export let KolTreeTag = 'kol-tree' as const;
export let KolTreeWcTag = 'kol-tree-wc' as const;
export let KolVersionTag = 'kol-version' as const;

export const setCustomTagNames = (transformTagName: (tagName: string) => string) => {
	KolAbbrTag = transformTagName(KolAbbrTag as string) as 'kol-abbr';
	KolAccordionTag = transformTagName(KolAccordionTag as string) as 'kol-accordion';
	KolAlertTag = transformTagName(KolAlertTag as string) as 'kol-alert';
	KolAlertWcTag = transformTagName(KolAlertWcTag as string) as 'kol-alert-wc';
	KolAvatarTag = transformTagName(KolAvatarTag as string) as 'kol-avatar';
	KolAvatarWcTag = transformTagName(KolAvatarWcTag as string) as 'kol-avatar-wc';
	KolBadgeTag = transformTagName(KolBadgeTag as string) as 'kol-badge';
	KolBreadcrumbTag = transformTagName(KolBreadcrumbTag as string) as 'kol-breadcrumb';
	KolButtonGroupTag = transformTagName(KolButtonGroupTag as string) as 'kol-button-group';
	KolButtonGroupWcTag = transformTagName(KolButtonGroupWcTag as string) as 'kol-button-group-wc';
	KolButtonLinkTag = transformTagName(KolButtonLinkTag as string) as 'kol-button-link';
	KolButtonTag = transformTagName(KolButtonTag as string) as 'kol-button';
	KolButtonWcTag = transformTagName(KolButtonWcTag as string) as 'kol-button-wc';
	KolCardTag = transformTagName(KolCardTag as string) as 'kol-card';
	KolDetailsTag = transformTagName(KolDetailsTag as string) as 'kol-details';
	KolDrawerTag = transformTagName(KolDrawerTag as string) as 'kol-drawer';
	KolFormTag = transformTagName(KolFormTag as string) as 'kol-form';
	KolHeadingTag = transformTagName(KolHeadingTag as string) as 'kol-heading';
	KolHeadingWcTag = transformTagName(KolHeadingWcTag as string) as 'kol-heading-wc';
	KolIconTag = transformTagName(KolIconTag as string) as 'kol-icon';
	KolImageTag = transformTagName(KolImageTag as string) as 'kol-image';
	KolIndentedTextTag = transformTagName(KolIndentedTextTag as string) as 'kol-indented-text';
	KolIndentedTextWcTag = transformTagName(KolIndentedTextWcTag as string) as 'kol-indented-text-wc';
	KolInputCheckboxTag = transformTagName(KolInputCheckboxTag as string) as 'kol-input-checkbox';
	KolInputColorTag = transformTagName(KolInputColorTag as string) as 'kol-input-color';
	KolInputDateTag = transformTagName(KolInputDateTag as string) as 'kol-input-date';
	KolInputEmailTag = transformTagName(KolInputEmailTag as string) as 'kol-input-email';
	KolInputFileTag = transformTagName(KolInputFileTag as string) as 'kol-input-file';
	KolInputNumberTag = transformTagName(KolInputNumberTag as string) as 'kol-input-number';
	KolInputPasswordTag = transformTagName(KolInputPasswordTag as string) as 'kol-input-password';
	KolInputRadioTag = transformTagName(KolInputRadioTag as string) as 'kol-input-radio';
	KolInputRangeTag = transformTagName(KolInputRangeTag as string) as 'kol-input-range';
	KolInputTag = transformTagName(KolInputTag as string) as 'kol-input';
	KolInputTextTag = transformTagName(KolInputTextTag as string) as 'kol-input-text';
	KolKolibriTag = transformTagName(KolKolibriTag as string) as 'kol-kolibri';
	KolLinkButtonTag = transformTagName(KolLinkButtonTag as string) as 'kol-link-button';
	KolLinkGroupTag = transformTagName(KolLinkGroupTag as string) as 'kol-link-group';
	KolLinkTag = transformTagName(KolLinkTag as string) as 'kol-link';
	KolLinkWcTag = transformTagName(KolLinkWcTag as string) as 'kol-link-wc';
	KolLogoTag = transformTagName(KolLogoTag as string) as 'kol-logo';
	KolModalTag = transformTagName(KolModalTag as string) as 'kol-modal';
	KolNavTag = transformTagName(KolNavTag as string) as 'kol-nav';
	KolPaginationTag = transformTagName(KolPaginationTag as string) as 'kol-pagination';
	KolPopoverWcTag = transformTagName(KolPopoverWcTag as string) as 'kol-popover-wc';
	KolProgressTag = transformTagName(KolProgressTag as string) as 'kol-progress';
	KolQuoteTag = transformTagName(KolQuoteTag as string) as 'kol-quote';
	KolSelectTag = transformTagName(KolSelectTag as string) as 'kol-select';
	KolSkipNavTag = transformTagName(KolSkipNavTag as string) as 'kol-skip-nav';
	KolSpanWcTag = transformTagName(KolSpanWcTag as string) as 'kol-span-wc';
	KolSpinTag = transformTagName(KolSpinTag as string) as 'kol-spin';
	KolSplitButtonTag = transformTagName(KolSplitButtonTag as string) as 'kol-split-button';
	KolSymbolTag = transformTagName(KolSymbolTag as string) as 'kol-symbol';
	KolTableStatefulTag = transformTagName(KolTableStatefulTag) as 'kol-table-stateful';
	KolTableStatelessTag = transformTagName(KolTableStatelessTag as string) as 'kol-table-stateless';
	KolTableStatelessWcTag = transformTagName(KolTableStatelessWcTag as string) as 'kol-table-stateless-wc';
	KolTableTag = transformTagName(KolTableTag as string) as 'kol-table';
	KolTabsTag = transformTagName(KolTabsTag as string) as 'kol-tabs';
	KolTextareaTag = transformTagName(KolTextareaTag as string) as 'kol-textarea';
	KolToastContainerTag = transformTagName(KolToastContainerTag as string) as 'kol-toast-container';
	KolToolbarTag = transformTagName(KolTooltipWcTag as string) as 'kol-toolbar';
	KolTooltipWcTag = transformTagName(KolTooltipWcTag as string) as 'kol-tooltip-wc';
	KolTreeItemTag = transformTagName(KolTreeItemTag as string) as 'kol-tree-item';
	KolTreeItemWcTag = transformTagName(KolTreeItemWcTag as string) as 'kol-tree-item-wc';
	KolTreeTag = transformTagName(KolTreeTag as string) as 'kol-tree';
	KolTreeWcTag = transformTagName(KolTreeWcTag as string) as 'kol-tree-wc';
	KolVersionTag = transformTagName(KolVersionTag as string) as 'kol-version';
};
