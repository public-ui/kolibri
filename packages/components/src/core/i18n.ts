import type { Generic } from 'adopted-style-sheets';
import locale_de from '../locales/de';
import locale_en from '../locales/en';

export type ResourcePrefix = 'Kol';
export type ComponentKeys = keyof typeof locale_de;

const mapLocaleKeys = (locale: { [K in ComponentKeys]: string }) =>
	(Object.keys(locale) as ComponentKeys[]).reduce((a, c) => ((a[`${'kol'}-${c}`] = locale[c]), a), {} as Generic.I18n.Map<ResourcePrefix, ComponentKeys>);

const TRANSLATIONS = new Set<Generic.I18n.RegisterPatch<Generic.I18n.Locale.ISO_639_1, ResourcePrefix, ComponentKeys>>([
	(t: (language: 'en', translationMap: Generic.I18n.Map<ResourcePrefix, ComponentKeys>) => Generic.I18n.Locale.ISO_639_1) => t('en', mapLocaleKeys(locale_en)),
	(t: (language: 'de', translationMap: Generic.I18n.Map<ResourcePrefix, ComponentKeys>) => Generic.I18n.Locale.ISO_639_1) => t('de', mapLocaleKeys(locale_de)),
]);

interface ITranslationOptions {
	/**
	 * Placeholders to insert into the text. Replacing {{key}} with the specified value if the "key".
	 */
	placeholders?: { [K: string]: string };
}

// String.replaceAll should be used but is only available with ECMAScript 2021.
const replaceAll = function (text: string, search: string, replacement: string) {
	return text.split(search).join(replacement);
};

type Resources = Map<string, string>;

export class I18nService {
	private resourceMap = new Map<Generic.I18n.Locale.ISO_639_1, Resources>();
	language: Generic.I18n.Locale.ISO_639_1;

	public constructor(initialLanguage?: Generic.I18n.Locale.ISO_639_1) {
		this.language = initialLanguage ?? 'de';
	}

	/**
	 * Adds a resource bundle for the specified language.
	 * @param lng the language the bundle is for
	 * @param translationMap the translations of the given language
	 */
	public addTranslations(
		translations?:
			| Generic.I18n.RegisterPatch<Generic.I18n.Locale.ISO_639_1, string, string>
			| Generic.I18n.RegisterPatch<Generic.I18n.Locale.ISO_639_1, string, string>[]
			| Set<Generic.I18n.RegisterPatch<Generic.I18n.Locale.ISO_639_1, string, string>>,
	) {
		if (Array.isArray(translations)) {
			translations = new Set(translations);
		} else if (typeof translations === 'function') {
			translations = new Set([translations]);
		}

		if (translations !== undefined) {
			translations.forEach((t) =>
				t((l, t) => {
					let resources = this.resourceMap.get(l);
					if (!resources) {
						resources = new Map<string, string>();
						this.resourceMap.set(l, resources);
					}

					Object.entries(t).forEach(([k, v]) => {
						if (typeof v === 'string') {
							resources.set(k, v);
						}
					});
					return l;
				}),
			);
		}
	}

	/**
	 * Set the current language.
	 * @param lng the language the bundle is for
	 */
	public setLanguage(lng: Generic.I18n.Locale.ISO_639_1) {
		this.language = lng;
	}

	/**
	 * Determines a human-readable translated text for the given resource key.
	 * @param key the resource key
	 * @param options optional translation parameters
	 * @returns the translated text
	 */
	public translate(key: string, options?: ITranslationOptions) {
		let result = this.resourceMap.get(this.language)?.get(key);
		if (result !== null) {
			if (options?.placeholders) {
				Object.entries(options.placeholders).forEach(([k, v]) => {
					result = replaceAll(result!, `{{${k}}}`, v);
				});
			}
		} else {
			result = key;
		}
		return result;
	}
}

let i18n: I18nService;

export const getI18nInstance = () => {
	return i18n;
};

export const initializeI18n = (
	lng: Generic.I18n.Locale.ISO_639_1,
	translations:
		| Generic.I18n.RegisterPatch<Generic.I18n.Locale.ISO_639_1, string, string>
		| Generic.I18n.RegisterPatch<Generic.I18n.Locale.ISO_639_1, string, string>[]
		| Set<Generic.I18n.RegisterPatch<Generic.I18n.Locale.ISO_639_1, string, string>> = TRANSLATIONS,
) => {
	i18n = new I18nService(lng);
	i18n.addTranslations(translations);
	return i18n;
};
