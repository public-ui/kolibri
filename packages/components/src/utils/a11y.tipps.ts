import { Log } from './dev.utils';

const a11yCache: Set<string> = new Set<string>();
export const a11yHint = (msg: string): void => {
	if (a11yCache.has(msg) === false) {
		a11yCache.add(msg);
		Log.debug(msg, {
			classifier: `âœ‹ a11y`,
			overwriteStyle: '; background-color: #09f',
		});
	}
};

const deprecatedCache: Set<string> = new Set<string>();
export const deprecatedHint = (msg: string): void => {
	if (deprecatedCache.has(msg) === false) {
		deprecatedCache.add(msg);
		Log.warn(msg, {
			classifier: `ðŸ”¥ deprecated`,
			forceLog: true,
			overwriteStyle: '; background-color: #f00',
		});
	}
};

const devCache: Set<string> = new Set<string>();
export const devHint = (msg: string): void => {
	if (devCache.has(msg) === false) {
		devCache.add(msg);
		Log.debug(msg, {
			classifier: `ðŸ’» dev`,
			overwriteStyle: '; background-color: #f09',
		});
	}
};
export const devWarning = (msg: string): void => {
	if (devCache.has(msg) === false) {
		devCache.add(msg);
		Log.warn(msg, {
			classifier: `ðŸ’» dev`,
			overwriteStyle: '; background-color: #f09',
		});
	}
};

const featureCache: Set<string> = new Set<string>();
export const featureHint = (msg: string, done = false): void => {
	if (featureCache.has(msg) === false) {
		featureCache.add(msg);
		msg += done === true ? ' âœ…' : '';
		Log.debug(msg, {
			classifier: `ðŸŒŸ feature`,
			overwriteStyle: '; background-color: #309',
		});
	}
};
devHint(
	`Wir freuen uns Ã¼ber jedes Feedback, Kommentare, Screenshots oder Demo-Links von einer auf KoliBri-basierenden Anwendung (kolibri@itzbund.de). Vielen Dank!`
);

const uiUxCache: Set<string> = new Set<string>();
export const uiUxHint = (msg: string): void => {
	if (uiUxCache.has(msg) === false) {
		uiUxCache.add(msg);
		Log.debug(msg, {
			classifier: `ðŸ“‘ ui/ux`,
			overwriteStyle: '; background-color: #060;',
		});
	}
};

export const a11yHintDisabled = (): void => {
	a11yHint(
		`"Disabled" schrÃ¤nkt die ZugÃ¤nglichkeit und Sichtbarkeit ein. Wir empfehlen aus Sicht der Barrierefreiheit readonly- statt disabled-Attribut zu verwenden.\n- https://uxdesign.cc/is-it-ok-to-grey-out-disabled-buttons-8afa74a0fae`
	);
};

export const a11yHintLabelingLandmarks = (value: unknown): void => {
	if (typeof value !== 'string' || value === '') {
		a11yHint(
			`Manche Strukturelemente, wie bspw. der nav-Tag, kÃ¶nnen mehrfach in einer Webseite verwendet werden. Damit die gleichnamigen Strukturelemente voneinander unterschieden werden kÃ¶nnen, ist es erforderlich ein Aria-Label zu setzen.\n- https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/Navigation_Role#accessibility_concerns`
		);
	}
};

export const uiUxHintMillerscheZahl = (className: string, length = 8): void => {
	if (length > 7) {
		uiUxHint(
			`[${className}] Innerhalb von Navigationsstrukturen wird empfohlen nicht mehr als 7 MenÃ¼punkte zu verwenden.

Links:
- https://de.ryte.com/wiki/Millersche_Zahl
- https://de.wikipedia.org/wiki/Millersche_Zahl`
		);
	}
};
