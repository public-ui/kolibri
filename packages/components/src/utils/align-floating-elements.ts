import { arrow, computePosition, flip, offset, shift } from '@floating-ui/dom';
import { processEnv } from '../schema';

import type { AlignPropType } from '../schema';
type Arguments = {
	floatingElement: HTMLElement;
	referenceElement: Element;
	arrowElement?: HTMLElement;
	align?: AlignPropType;
};
export const alignFloatingElements = async ({ floatingElement, referenceElement, arrowElement, align = 'top' }: Arguments) => {
	if (processEnv !== 'test') {
		const middleware = [offset(arrowElement?.offsetHeight ?? 10), flip(), shift()];
		if (arrowElement) {
			middleware.push(arrow({ element: arrowElement }));
		}

		const { x, y, middlewareData, placement } = await computePosition(referenceElement, floatingElement, {
			placement: align,
			middleware: middleware,
			strategy: 'fixed',
		});

		Object.assign(floatingElement.style, {
			left: `${x}px`,
			top: `${y}px`,
			visibility: 'visible',
		});

		if (arrowElement) {
			if (typeof middlewareData.arrow?.x === 'number') {
				Object.assign(arrowElement.style, {
					left: `${middlewareData.arrow.x}px`,
					top: placement === 'bottom' ? `${-arrowElement.offsetHeight / 2}px` : '',
					bottom: placement === 'top' ? `${-arrowElement.offsetHeight / 2}px` : '',
				});
			} else if (typeof middlewareData.arrow?.y === 'number') {
				Object.assign(arrowElement.style, {
					left: placement === 'right' ? `${-arrowElement.offsetWidth / 2}px` : '',
					right: placement === 'left' ? `${-arrowElement.offsetWidth / 2}px` : '',
					top: `${middlewareData.arrow.y}px`,
				});
			}
		}
	}
};
