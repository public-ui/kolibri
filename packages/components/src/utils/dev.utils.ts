import { ModalService } from '../components/modal/service';
import { processEnv } from './reuse';

let WINDOW: Window | null = null;
let DOCUMENT: Document | null = null;

export const configKoliBri = (window: Window): void => {
	if (window instanceof Window) {
		WINDOW = window;
		if (WINDOW.document instanceof Document) {
			DOCUMENT = window.document;
		} else {
			console.warn(`The given Window has no valid Document.`);
		}
	} else {
		console.warn(`The given Window is not valid.`);
	}
};

export const getWindow = (): Window => (WINDOW || typeof window === 'undefined' ? (null as unknown as Window) : window);
export const getDocument = (): Document => (DOCUMENT || typeof getWindow().document === 'undefined' ? (null as unknown as Document) : getWindow().document);

let META_CONFIG: string | null = null;
let DEV_MODE: boolean | null = null;
let EXPERIMENTAL_MODE: boolean | null = null;
let COLOR_CONTRAST_ANALYSIS: boolean | null = null;

export const getDevMode = (): boolean => DEV_MODE === true;
export const getExperimentalMode = (): boolean => EXPERIMENTAL_MODE === true;
export const getColorContrastAnalysis = (): boolean => COLOR_CONTRAST_ANALYSIS === true;

type LogShield = {
	label: string;
	style: string;
};

type LogShieldOptions = {
	classifier?: string;
	forceLog?: boolean;
	overwriteStyle?: string;
};

export class Log {
	private static shield: LogShield = {
		label: '%cKoliBri',
		style: 'color: white; background: #666; font-weight: bold; padding: .25em .5em; border-radius: 3px; border: 1px solid #000',
	};

	private static mapToArray(msg: unknown | unknown[]): unknown[] {
		return Array.isArray(msg) ? msg : [msg];
	}

	private static handleClassifier(classifier?: string): string {
		if (typeof classifier === 'string' && classifier.length > 0) {
			return `${Log.shield.label} | ${classifier}`;
		} else {
			return Log.shield.label;
		}
	}

	private static getShield(options?: LogShieldOptions): string[] {
		return [Log.handleClassifier(options?.classifier), `${Log.shield.style};${options?.overwriteStyle || ''}`];
	}

	public static debug(msg: unknown | unknown[], options?: LogShieldOptions): void {
		if (DEV_MODE || options?.forceLog === true) {
			console.debug(...Log.getShield(options), ...Log.mapToArray(msg));
		}
	}

	public static info(msg: unknown | unknown[], options?: LogShieldOptions): void {
		if (DEV_MODE || options?.forceLog === true) {
			console.info(...Log.getShield(options), ...Log.mapToArray(msg));
		}
	}

	public static trace(msg: unknown | unknown[], options?: LogShieldOptions): void {
		if (DEV_MODE || options?.forceLog === true) {
			console.trace(...Log.getShield(options), ...Log.mapToArray(msg));
		}
	}

	public static warn(msg: unknown | unknown[], options?: LogShieldOptions): void {
		if (DEV_MODE || options?.forceLog === true) {
			console.warn(...Log.getShield(options), ...Log.mapToArray(msg));
		}
	}

	public static error(msg: unknown | unknown[], options?: LogShieldOptions): void {
		if (DEV_MODE || options?.forceLog === true) {
			console.error(...Log.getShield(options), ...Log.mapToArray(msg));
		}
	}

	public static throw(msg: unknown | unknown[], options?: LogShieldOptions): void {
		if (DEV_MODE || options?.forceLog === true) {
			throw new Error(...Log.getShield(options), ...Log.mapToArray(msg));
		}
	}
}

const initMeta = (): void => {
	if (DEV_MODE === null && EXPERIMENTAL_MODE === null && COLOR_CONTRAST_ANALYSIS === null) {
		const meta = getDocument().querySelector('meta[name="kolibri"]');
		if (meta && meta.hasAttribute('content')) {
			META_CONFIG = meta.getAttribute('content');
			if (typeof META_CONFIG === 'string') {
				DEV_MODE = META_CONFIG.includes('dev-mode=true');
				EXPERIMENTAL_MODE = META_CONFIG.includes('experimental-mode=true');
				COLOR_CONTRAST_ANALYSIS = META_CONFIG.includes('color-contrast-analysis=true');
			}
		}
	} else {
		console.warn(`You can only initialize DEV_MODE and COLOR_CONTRAST_ANALYSIS once.`);
	}
};

const getKoliBri = (): Record<string, unknown> => {
	let kolibri = getWindow().KoliBri;
	if (kolibri === undefined) {
		kolibri = {};
		Object.defineProperty(getWindow(), 'KoliBri', {
			value: kolibri,
			writable: false,
		});
	}
	return kolibri;
};

export const initKoliBri = (): void => {
	if (getKoliBri().Modal === undefined) {
		const Modal = new ModalService();
		Object.defineProperty(getKoliBri(), 'Modal', {
			get: function (): ModalService {
				return Modal;
			},
		});
		initMeta();
		Log.debug(
			`
	,--. ,--.         ,--. ,--. ,-----.           ,--.
	|  .'   /  ,---.  |  | \`--' |  |) /_  ,--.--. \`--'
	|  .   '  | .-. | |  | ,--. |  .-.  \\ |  .--' ,--.
	|  |\\   \\ | '-' | |  | |  | |  '--' / |  |    |  |
	\`--' \`--Â´  \`---Â´  \`--' \`--' \`------Â´  \`--'    \`--'
	ðŸš¹ The accessible HTML-Standard | ðŸ‘‰ https://public-ui.github.io | 1.7.9
		`,
			{
				forceLog: true,
			}
		);
	} else {
		console.warn(`You can only initialize KoliBri once.`);
	}
};
export { getKoliBri };

export const renderDevAdvice = (): void => {
	if (getKoliBri().adviceShown !== true) {
		Object.defineProperty(getKoliBri(), 'adviceShown', {
			get: function () {
				return true;
			},
		});
		Log.debug(
			`
You are using the KoliBri component library. If you have any suggestions for improvement or find a problem, please contact us:

Ticket: https://github.com/public-ui/kolibri/issues/new/choose (for privacy reasons, please use email)
Email: kolibri@itzbund.de
`
		);
	}
};

let nonce = (): string => Math.floor(Math.random() * 16777215).toString(16);

if (processEnv === 'test') {
	nonce = (): string => 'nonce';
}

export { nonce };
