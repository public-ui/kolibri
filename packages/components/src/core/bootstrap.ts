import type { Generic, LoaderCallback, RegisterOptions } from 'adopted-style-sheets';
import { register as coreRegister } from 'adopted-style-sheets';
import { configI18n, initI18n } from './i18n';

type KoliBriOptions = RegisterOptions & {
	// transformTagName?: (tagName: string) => string;
};

export const bootstrap = async (
	themes:
		| Generic.Theming.RegisterPatch<string, string, string>
		| Generic.Theming.RegisterPatch<string, string, string>[]
		| Set<Generic.Theming.RegisterPatch<string, string, string>>,
	loaders: LoaderCallback | LoaderCallback[] | Set<LoaderCallback>,
	options?: KoliBriOptions
): Promise<void[]> => {
	await initI18n(options?.translation?.name);
	await configI18n(options?.translation?.name ?? 'de', options?.translations);
	// configTransformTagName(options.transformTagName);
	return await coreRegister(themes, loaders, options);
};

export const register = bootstrap;
