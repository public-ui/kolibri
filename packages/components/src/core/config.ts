import type { RegisterOptions } from 'adopted-style-sheets';
import { configI18n } from './i18n';

type KoliBriConfigOptions = Omit<RegisterOptions, 'theme'>;

export const configKoliBri = async (options: KoliBriConfigOptions): Promise<void> => {
	await configI18n(options?.translation?.name ?? 'de', options?.translations);
};
