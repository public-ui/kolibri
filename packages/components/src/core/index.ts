import type { Generic, LoaderCallback, RegisterOptions } from 'adopted-style-sheets';
import { register as coreRegister } from 'adopted-style-sheets';
import { configKoliBri } from './config';

export const register = async (
	themes:
		| Generic.Theming.RegisterPatch<string, string, string>
		| Generic.Theming.RegisterPatch<string, string, string>[]
		| Set<Generic.Theming.RegisterPatch<string, string, string>>,
	loaders: LoaderCallback | LoaderCallback[] | Set<LoaderCallback>,
	options?: RegisterOptions
): Promise<void[]> => {
	await configKoliBri({
		translation: options?.translation,
		translations: options?.translations,
	});
	return await coreRegister(themes, loaders, options);
};
