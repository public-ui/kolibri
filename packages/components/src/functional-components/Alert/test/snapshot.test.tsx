import { h } from '@stencil/core';
import KolAlertFc from './..';
import { renderFunctionalComponentToSpecPage } from '../../../utils/testing';

describe('KolAlertFc', () => {
	it('should render', async () => {
		const page = await renderFunctionalComponentToSpecPage(() => <KolAlertFc />);

		expect(page.root).toMatchSnapshot();
	});
});
