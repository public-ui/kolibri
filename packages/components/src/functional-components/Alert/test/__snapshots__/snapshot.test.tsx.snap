// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`KolAlertFc should render 1`] = `
<div class="alert default kol-alert-wc msg">
  <div class="heading">
    <kol-icon _icons="codicon codicon-comment" _label="kol-message" class="heading-icon"></kol-icon>
    <div class="heading-content">
      <div class="content"></div>
    </div>
  </div>
</div>
`;
